AG-9272 Error Bar mouse hover implementation

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-9272

This PR:
- Highlights the datum when the mouse hovers over an error bars scene node.
- Shows the tooltip when the mouse hovers over an error bars scene node.
- Renders the tooltip higher up to avoid covering the error bar.

Notes:
- Interaction range `'exact'` and `'nearest'` are treated the same as `range = 0`
- `ErrorBarNode` uses a hierarchical bounding box structure for precise & high-performance hit-testing.
- Tested on https://plnkr.co/edit/p9mib2yv6YYY97uL?open=main.js

**Before:**
<img width="235" alt="Screenshot 2023-10-20 at 14 44 29" src="https://github.com/ag-grid/ag-charts/assets/23700103/24d9e478-56ff-49f4-8c25-21148122473f">

**After:**
<img width="232" alt="Screenshot 2023-10-20 at 14 44 49" src="https://github.com/ag-grid/ag-charts/assets/23700103/473fa0ce-666f-41c9-8054-ce300fa15555">

olegat committed:
AG-8238 Refactor how errorBar.ts uses datums

The error bar points data is now part of the ErrorBoundSeriesNodeDatum
type from the community package. This avoids the needs for managing two
`nodeData` variables.

olegat committed:
AG-9272 Implement ErrorBars mouse hover hit testing

This correctly applies the highlighting style when the mouse hovers over
an error bar. However, this doesn't show the tooltip yet.

olegat committed:
AG-9272 Don't render tooltip over error bar

On line and scatter series, using datum.midPoint renders the tooltip
over the top part of the error bars. So use datum.yBar.upperPoint
instead if that property exists.

olegat committed:
AG-9272 Show tooltip when mouse hovers on errorbars

Move some of the functionality of chart.ts into a utility method in
tooltipManager.ts so that the error-bar module can also show tooltips.

olegat commented on packages/ag-charts-community/src/chart/interaction/tooltipManager.ts:
```diff
@@ -106,4 +110,46 @@ export class TooltipManager {
         this.appliedState = { content: contentToApply, meta: metaToApply };
     }
+
+    public static makeTooltipMeta(
```
Should this static method be moved into its own file (e.g. _tooltipUtils.ts_)?

The new imports are creating a lot of warnings in `nx run ag-charts-community:lint-depcruise` (although this should be okay because it's only `import type` statements).

alantreadway replied:
So I don't think moving this method to a new file will help with the circular deps; the problem is the dependency back to `SeriesNodeDatum` and `ErrorBoundSeriesNodeDatum ` - we need to either not use these, or break them into their own minimally dependent modules?

I'd prefer if we resolve this now if straightforward.

olegat replied:
Okay. I think that moving `SeriesNodeDatum` and `ErrorBoundSeriesNodeDatum` is the preferred option. So I'll see whether that's viable.

Not using them (i.e. redefining the type contract in _tooltipManager.ts_) will be our Plan B.

olegat replied:
I've moved these types into _seriesTypes.ts_ just below the `ISeries` interface. This seems like a decent solution. Please let me know your thoughts.

olegat commented on packages/ag-charts-community/src/chart/interaction/tooltipManager.ts:
```diff
@@ -106,4 +110,46 @@ export class TooltipManager {
+
+        // On line and scatter series, the tooltip covers the top of errorbars when using
+        // datum.midPoint. Using datum.yBar.upperPoint renders the tooltip higher up.
+        const refPoint: Point | undefined = datum.yBar?.upperPoint ?? datum.midPoint;
```
Note: see PR description for a preview of what the output looks like with `midPoint` and `upperPoint`

olegat committed:
AG-9272 Implement tooltip.range=number on error bars

The highlight logic can start spasming when multiple nodes match the
hit testing. We may want to discuss how to handle this.

alantreadway commented on packages/ag-charts-enterprise/src/features/error-bar/errorBarNode.ts:
```diff
@@ -85,5 +95,49 @@ export class ErrorBarNode extends _Scene.Group {
+                }
+            }
+        }
+        return false;
```
Nit: Could we reduce the nesting here by short-circuiting the outer condition?

e.g.:
```
if (!this.bboxes.union.containsPoint(x, y)) return false;
```

olegat replied:
Sure, happy to make this change. I generally believe that having few `return` statements makes the code more readable, but I don't think that it matters much in this case.

olegat replied:
Done

olegat committed:
AG-9272 Move datum types to seriesTypes.ts

This avoids too many cyclical imports in tooltipManager.ts (which uses
the SeriesNodeDatum and ErrorBoardSeriesNodeDatum in the makeTooltipMeta
method).

olegat committed:
AG-9272 Short-circuit errorBarNode hit testing

alantreadway reviewed (approved):
Nice, thanks for the follow-up!
